samples/sample_multi_transcode/src/sample_multi_transcode.cpp

```diff
@@ -358,90 +359,98 @@ void Launcher::DoTranscoding()
+                        msdk_printf(MSDK_STRING("%s"), ss.str().c_str());
+
+                        for (auto context : m_pThreadContextArray)
                         {
-                            msdk_stringstream ss;
-                            ss << MSDK_STRING("\n\n session ") << i << MSDK_STRING(" [") << m_pSessionArray[i]->pPipeline->GetSessionText() << MSDK_STRING("] failed with status ") << StatusToString(m_pSessionArray[i]->transcodingSts)
-                               << MSDK_STRING(" shutting down the application...") << std::endl << std::endl;
-                            msdk_printf(MSDK_STRING("%s"), ss.str().c_str());
-
-                            for (size_t j = 0; j < m_pSessionArray.size(); j++)
-                            {
-                                m_pSessionArray[j]->pPipeline->StopSession();
-                            }
+                            context->pPipeline->StopSession();
                         }
                     }
-
-                    // Now clear its handle and thread info
-                    MSDK_SAFE_DELETE(m_HDLArray[i]);
-                    m_HDLArray[i] = NULL;
+                }
+                else if (m_pThreadContextArray[i]->transcodingSts > MFX_ERR_NONE)
+                {
+                    msdk_stringstream ss;
+                    ss << MSDK_STRING("\n\n session ") << i << MSDK_STRING(" [")
+                    << m_pThreadContextArray[i]->pPipeline->GetSessionText()
+                    << MSDK_STRING("] returned warning status ")
+                    << StatusToString(m_pThreadContextArray[i]->transcodingSts)
+                    << std::endl << std::endl;
+                    msdk_printf(MSDK_STRING("%s"), ss.str().c_str());
+                }
+            }
+            else
+            {
+                aliveSessionsCount++;
+                if (m_pThreadContextArray[i]->pPipeline->IsOverlayUsed()) {
+                    aliveOverlaySessionsCount++;
                 }
             }
         }
 
         // If all sessions are already stopped (no matter with error or not) - we need to forcibly stop all overlay sessions
-        if (aliveSessionsCount <= nOverlayThreads + 1)
+        if (aliveOverlaySessionsCount && aliveSessionsCount <= aliveOverlaySessionsCount + 1)
         {
             // Sending stop message
-            for (size_t i = 0; i < totalSessions; i++)
+            for (auto context : m_pThreadContextArray)
             {
-                if (m_HDLArray[i] && m_pSessionArray[i]->pPipeline->IsOverlayUsed())
+                if (!context->handle.valid())
+                    continue;
+
+                auto waitSts = context->handle.wait_for(std::chrono::milliseconds(1));
+                if (waitSts != std::future_status::ready &&
+                    context->pPipeline->IsOverlayUsed())
                 {
-                    m_pSessionArray[i]->pPipeline->StopSession();
+                    context->pPipeline->StopSession();
```

Comment:
@vilichev could you check the following thing, we already sent Stop to all sessions at line 403 (no matter overlay or not). Why do we need to do it once more? Looks like there is no need for it

Reply:
As we agreed, we need to forcibly stop of overlay sessions since they are never stopped by themselves. Thanks for your help @mgonchar !